Added GMMConv

melioristic opened the pull request:
Added GMMConv from the paper: Geometric deep learning on graphs and manifolds using mixture model CNNs

CarloLucibello commented on src/layers/conv.jl:
````diff
@@ -1065,4 +1065,117 @@ function (l::MEGNetConv)(g::GNNGraph, x::AbstractMatrix, e::AbstractMatrix)
+
+where
+```math
+\mathbf{w_{k}} = \exp(\frac{-1}{2}(u-\mu_k)^T \Sigma_k^{-1}(u-\mu
+_k))
````
```suggestion
\mathbf{w_{k}} = \exp(\frac{-1}{2}(u-\mu_k)^T \Sigma_k^{-1}(u-\mu_k))
```
Also, \mu and \Sigma need to be specified

CarloLucibello replied:
no actually is not a scalar

CarloLucibello replied:
Ok, I edited the suggestion so that now it should be mathematically clearer what's going on

melioristic replied:
Thanks!

CarloLucibello commented on src/layers/conv.jl:
````diff
@@ -1065,4 +1065,117 @@ function (l::MEGNetConv)(g::GNNGraph, x::AbstractMatrix, e::AbstractMatrix)
+
+Performs the operation
+```math
+\mathbf{x}_i' = \frac{1}{|N(i)|} \sum_{j\in N(i)}\frac{1}{K}\sum_{k=1}^k\mathbf{w}_k(u_{i,j})\odot \mathbf{\theta}_k \mathbf{x}_j
````
```suggestion
\mathbf{x}_i' = \frac{1}{|N(i)|} \sum_{j\in N(i)}\frac{1}{K}\sum_{k=1}^k w_k(u_{j\to i}) \odot \mathbf{\theta}_k \mathbf{x}_j
```

CarloLucibello commented on src/layers/conv.jl:
```diff
@@ -1065,4 +1065,117 @@ function (l::MEGNetConv)(g::GNNGraph, x::AbstractMatrix, e::AbstractMatrix)
+    GMMConv(mu, sigma_inv, b, σ, ch, n_kernel, u_dim, dense_x)
+end
 
+function (l::GMMConv)(g::GNNGraph, x::AbstractMatrix, u::AbstractMatrix)
```
let's rename `u` to `e` since we usually call by that the edge features

melioristic replied:
Will do that in the next commit

CarloLucibello commented on src/layers/conv.jl:
```diff
@@ -1065,4 +1065,117 @@ function (l::MEGNetConv)(g::GNNGraph, x::AbstractMatrix, e::AbstractMatrix)
 
+function (l::GMMConv)(g::GNNGraph, x::AbstractMatrix, u::AbstractMatrix)
+
+    if g.graph[3]!==nothing
+        @warn g.graph[3] !== nothing "Edge weights $(g.graph[3]) provided, will not be used for computation"
+    end
```
```suggestion
```
we can remove this, weight are not used unless explicitly stated, no need to warn

CarloLucibello commented on src/layers/conv.jl:
```diff
@@ -1065,4 +1065,117 @@ function (l::MEGNetConv)(g::GNNGraph, x::AbstractMatrix, e::AbstractMatrix)
+
+    num_edges = g.num_edges
+    d = degree(g, dir=:in)
+    u = reshape(u,(num_edges, 1, l.u_dim))
```
```suggestion
    u = reshape(u, (num_edges, 1, l.u_dim))
```

CarloLucibello replied:
The reshape is wrong actually. While in python edge features are of size `num_edges x dim`, for us it is `dim x num_edges`. So you should reshape 
```julia
u = reshape(u, (l.u_dim, 1, l.num_edges))
mu = reshape(mu, (l.u_dim, l.n_kernel, 1))
```

CarloLucibello commented on src/layers/conv.jl:
```diff
@@ -1065,4 +1065,117 @@ function (l::MEGNetConv)(g::GNNGraph, x::AbstractMatrix, e::AbstractMatrix)
+    w = -0.5*(u.-mu).^2
+    w = w .* ((reshape(l.sigma_inv, (1, l.n_kernel, l.u_dim)).^2) )
+    w = exp.(sum(w, dims = 3 )) # n_edges, n_kernel,  1
+    w = permutedims(w, [3,2,1])
```
ideally if reshape is done properly there should be no need for permutation

CarloLucibello commented on src/layers/conv.jl:
```diff
@@ -1065,4 +1065,117 @@ function (l::MEGNetConv)(g::GNNGraph, x::AbstractMatrix, e::AbstractMatrix)
+    
+    x = propagate(e_mul_xj, g, +, xj=xj, e=w)
+    x = dropdims(mean(x, dims=2), dims=2)
+    x = 1/d.*x
```
```suggestion
    x = 1 / d .* x
```

CarloLucibello commented:
Tests are missing

melioristic commented:
Added the changes.
- Added the learnable parameter in the doc
- Changed the name from u to e in view of the package convention 
- As a convention, [last dimension as num_edges.](https://github.com/CarloLucibello/GraphNeuralNetworks.jl/pull/147#discussion_r831118362)

Still need to add the test.

CarloLucibello commented on src/layers/conv.jl:
```diff
@@ -1065,4 +1065,103 @@ function (l::MEGNetConv)(g::GNNGraph, x::AbstractMatrix, e::AbstractMatrix)
+g = GNNGraph(s,t)
+in_feature, out_feature, n_k, e_dim = 4, 7, 8, 10
+x = randn(in_feature, g.num_nodes)
+u = randn(e_dim, g.num_edges)
+
+# create layer
+l = GMMConv(in_feature=>out_feature, n_k, e_dim)
+
+# forward pass
+l(g, x, u)
```
```suggestion
e = randn(e_dim, g.num_edges)

# create layer
l = GMMConv(in_feature=>out_feature, n_k, e_dim)

# forward pass
l(g, x, e)
```

CarloLucibello commented on src/layers/conv.jl:
```diff
@@ -1065,4 +1065,103 @@ function (l::MEGNetConv)(g::GNNGraph, x::AbstractMatrix, e::AbstractMatrix)
+    in, out, n_kernel, e_dim = l.ch[1], l.ch[2], l.n_kernel, l.e_dim
+    print(io, "GMMConv(", in, " => ", out)
+    print(io, ", n_kernel= ", n_kernel)
+    print(io, ", pseudo-cordinate dimension = ", e_dim)
```
```suggestion
    print(io, ", e_dim = ", e_dim)
```

melioristic commented on src/layers/conv.jl:
```diff
@@ -1065,4 +1065,103 @@ function (l::MEGNetConv)(g::GNNGraph, x::AbstractMatrix, e::AbstractMatrix)
+
+Flux.@functor GMMConv
+
+function GMMConv(ch::Pair{Int, Int}, 
+                n_kernel::Int,
+                e_dim::Int,
+                σ=identity;
```
Is it a good idea to make e_dim default to 1? People might unknowingly try to use e of different dimensions and see errors.

melioristic commented:
> 

Test added as well

CarloLucibello commented on test/layers/conv.jl:
```diff
@@ -262,4 +262,14 @@
+        l = GMMConv(in_channel => out_channel, K=K, e_dim=e_dim)
+        for g in test_graphs
+            g = GNNGraph(g, edata=rand(e_dim, g.num_edges))
+            test_layer(l, g, rtol=RTOL_HIGH, outsize = (out_channel, g.num_edges))
```
```suggestion
            test_layer(l, g, rtol=RTOL_HIGH, outsize = (out_channel, g.num_nodes))
```

melioristic replied:
Ahh My bad!

CarloLucibello commented on src/layers/conv.jl:
```diff
@@ -1065,4 +1065,102 @@ function (l::MEGNetConv)(g::GNNGraph, x::AbstractMatrix, e::AbstractMatrix)
 end
 
+@doc raw"""
+    GMMConv(in => out, K, e_dim, σ=identity; [init, bias])
+Graph mixture model convolution layer from the paper [Geometric deep learning on graphs and manifolds using mixture model CNNs](https://arxiv.org/abs/1611.08402)
```
```suggestion
    GMMConv(in => out, e_dim, σ=identity; [K, init, bias])

Graph mixture model convolution layer from the paper [Geometric deep learning on graphs and manifolds using mixture model CNNs](https://arxiv.org/abs/1611.08402).
```

CarloLucibello replied:
Maybe the best compromise is to have `e_dim` has a mandatory positional argument and `K`  as keyword argument with default 1. Or we could have `e_dim` *mandatory* keyword arg. What do you think?  I don't want both of them to be positional since something like `GMMConv(16 => 16, 3, 5)`  is very unclear.

CarloLucibello replied:
Actually, we should adopt the interface provided by `CGConv`  and have `GMMConv((in, ein) => out, act=identity; K=1, bias=true, init=glorot_uniform, residual=false)`.

melioristic replied:
Working on it.

melioristic replied:
Added in the [commit](https://github.com/melioristic/GraphNeuralNetworks.jl/commit/510e4309b36e064ac42c441429e97b6990313d1d)

CarloLucibello commented on src/layers/conv.jl:
```diff
@@ -1065,4 +1065,102 @@ function (l::MEGNetConv)(g::GNNGraph, x::AbstractMatrix, e::AbstractMatrix)
+# Arguments 
+- `in`: Number of input features.
+- `out`: Number of output features.
+- `K` : Number of kernels.
```
```suggestion
- `K` : Number of kernels (Default `1`).
```

CarloLucibello commented on src/layers/conv.jl:
```diff
@@ -1065,4 +1065,102 @@ function (l::MEGNetConv)(g::GNNGraph, x::AbstractMatrix, e::AbstractMatrix)
+- `in`: Number of input features.
+- `out`: Number of output features.
+- `K` : Number of kernels.
+- `e_dim` : Dimensionality of pseudo coordinates.
```
```suggestion
- `e_dim` : Dimensionality of pseudo coordinates. Has to correspond to the edge features dimension in the forward pass.
```

CarloLucibello commented on src/layers/conv.jl:
````diff
@@ -1065,4 +1065,102 @@ function (l::MEGNetConv)(g::GNNGraph, x::AbstractMatrix, e::AbstractMatrix)
+l(g, x, e)
+```
+"""
+
````
```suggestion
```
Otherwise the string won't be taken as a docstring

CarloLucibello commented on src/layers/conv.jl:
```diff
@@ -1065,4 +1065,102 @@ function (l::MEGNetConv)(g::GNNGraph, x::AbstractMatrix, e::AbstractMatrix)
+
+    num_edges = g.num_edges
+    d = degree(g, dir=:in)
+    u = reshape(u, (l.e_dim, 1, num_edges))
```
Let's respect the notation in the docstring formula. This should be called `e`, and what is currently called `e` should be called `w`.

CarloLucibello commented on src/layers/conv.jl:
```diff
@@ -1065,4 +1065,101 @@ function (l::MEGNetConv)(g::GNNGraph, x::AbstractMatrix, e::AbstractMatrix)
+    num_edges = g.num_edges
+    d = degree(g, dir=:in)
+    u = reshape(u, (l.e_dim, 1, num_edges))
+    mu = reshape(l.mu, (l.e_dim, l.K, 1))
```
mu is created as a `K x edim`, so the reshape should be `K x edim x 1`  to preserve natural ordering. Or mu should be created `edim x K`.

CarloLucibello commented on src/layers/conv.jl:
```diff
@@ -1065,4 +1065,101 @@ function (l::MEGNetConv)(g::GNNGraph, x::AbstractMatrix, e::AbstractMatrix)
+    mu = reshape(l.mu, (l.e_dim, l.K, 1))
+    
+    e = -0.5*(u.-mu).^2
+    e = e .* ((reshape(l.sigma_inv, (l.e_dim, l.K, 1)).^2) )
```
```suggestion
    e = e .* reshape(l.sigma_inv, (l.e_dim, l.K, 1))
```
No need to square since `sigma_inv`  is an inverse covariance

CarloLucibello commented on src/layers/conv.jl:
```diff
@@ -1065,4 +1065,101 @@ function (l::MEGNetConv)(g::GNNGraph, x::AbstractMatrix, e::AbstractMatrix)
+    e = e .* ((reshape(l.sigma_inv, (l.e_dim, l.K, 1)).^2) )
+    e = exp.(sum(e, dims = 1 )) # (1, K, num_edge) 
+
+    xj = reshape(l.dense_x(x), (l.ch[2],l.K,:)) # (out, K, num_nodes) 
```
```suggestion
    xj = reshape(l.dense_x(x), (l.ch[2], l.K, :)) # (out, K, num_nodes) 
```

melioristic replied:
I did not get this part, I think it should still be K. Don't we need some transformation of x for each kernel K.

CarloLucibello replied:
yes, you are right, ignore that

melioristic replied:
Thanks!

melioristic commented:
Added 
- (in, ein)=>out, with no default for ein, so the layer cannot be instantiated with GMMConv(in=>out), would prefer this way, but can discuss
- Added residual connection 
- Changed test accordingly
- Changed (cosmetic) the equation for weights, hope it looks understandable when rendered.

CarloLucibello commented on src/layers/conv.jl:
```diff
@@ -1065,4 +1065,116 @@ function (l::MEGNetConv)(g::GNNGraph, x::AbstractMatrix, e::AbstractMatrix)
 end
 
+@doc raw"""
+    GMMConv((in, ein) => out, σ=identity; K=1, bias=true, init=glorot_uniform, residual=false)
```
```suggestion
    GMMConv((in, ein) => out, σ=identity; K=1, bias=true, init=glorot_uniform, residual=false)

```

CarloLucibello replied:
newline after signature

CarloLucibello commented on src/layers/conv.jl:
```diff
@@ -1065,4 +1065,116 @@ function (l::MEGNetConv)(g::GNNGraph, x::AbstractMatrix, e::AbstractMatrix)
+- `K`: Number of kernels. Default `1`.
+- `bias`: Add learnable bias. Default `true`.
+- `init`: Weights' initializer. Default `glorot_uniform`.
+- `residual`: Residual conncetion. Default `false`
```
```suggestion
- `residual`: Residual conncetion. Default `false`.
```

CarloLucibello commented on src/layers/conv.jl:
```diff
@@ -1065,4 +1065,116 @@ function (l::MEGNetConv)(g::GNNGraph, x::AbstractMatrix, e::AbstractMatrix)
+    
+    (nin, ein), out = ch
+    mu = init(ein, K)
+    sigma_inv = init(K, ein)
```
```suggestion
    sigma_inv = init(ein, K)
```

CarloLucibello commented on src/layers/conv.jl:
```diff
@@ -1065,4 +1065,116 @@ function (l::MEGNetConv)(g::GNNGraph, x::AbstractMatrix, e::AbstractMatrix)
+    w = reshape(e, (ein, 1, num_edges))
+    mu = reshape(l.mu, (ein, l.K, 1))
+    
+    w = -0.5*(w.-mu).^2
```
```suggestion
    w = -0.5 * (w .- mu).^2
```

CarloLucibello commented on src/layers/conv.jl:
```diff
@@ -1065,4 +1065,116 @@ function (l::MEGNetConv)(g::GNNGraph, x::AbstractMatrix, e::AbstractMatrix)
+    w = w .* reshape(l.sigma_inv, (ein, l.K, 1))
+    w = exp.(sum(w, dims = 1 )) # (1, K, num_edge) 
+
+    xj = reshape(l.dense_x(x), (out,l.K,:)) # (out, K, num_nodes) 
```
```suggestion
    xj = reshape(l.dense_x(x), (out, l.K, :)) # (out, K, num_nodes) 
```

CarloLucibello commented on src/layers/conv.jl:
```diff
@@ -1065,4 +1065,116 @@ function (l::MEGNetConv)(g::GNNGraph, x::AbstractMatrix, e::AbstractMatrix)
+    m = l.σ(m .+ l.bias)
+    
+    if l.residual
+        if size(x,1)==size(m,1)
```
```suggestion
        if size(x, 1) == size(m, 1)
```

CarloLucibello commented on src/layers/conv.jl:
```diff
@@ -1065,4 +1065,116 @@ function (l::MEGNetConv)(g::GNNGraph, x::AbstractMatrix, e::AbstractMatrix)
+    
+    if l.residual
+        if size(x,1)==size(m,1)
+            m+=x
```
```suggestion
            m += x
```

CarloLucibello commented:
mostly stile comments, almost ready to go

CarloLucibello commented on src/layers/conv.jl:
```diff
@@ -1065,4 +1065,117 @@ function (l::MEGNetConv)(g::GNNGraph, x::AbstractMatrix, e::AbstractMatrix)
+function (l::GMMConv)(g::GNNGraph, x::AbstractMatrix, e::AbstractMatrix)
+    (nin, ein), out = l.ch #Notational Simplicity
+
+    @assert (ein == size(e)[1] && g.num_edges == size(e)[2]) "Pseudo-cordinate dim $(size(u)) does not match (ein=$(ein),num_edge=$(g.num_edges))"
```
```suggestion
    @assert (ein == size(e)[1] && g.num_edges == size(e)[2]) "Pseudo-cordinate dim $(size(e)) does not match (ein=$(ein),num_edge=$(g.num_edges))"
```

melioristic commented:
> mostly stile comments, almost ready to go

Just being curious, do you use some kind of automatic formatter, any suggestions there.

CarloLucibello commented:
A first test failure could be solved by adding 
```
(l::GMMConv)(g::GNNGraph) = GNNGraph(g, ndata=l(g, node_features(g), edge_features(g)))
```

CarloLucibello commented on test/layers/conv.jl:
```diff
@@ -262,4 +262,14 @@
+        K = 5
+        l = GMMConv((in_channel, ein_channel )=> out_channel, K=K)
+        for g in test_graphs
+            g = GNNGraph(g, edata=rand(ein_channel, g.num_edges))
```
```suggestion
            g = GNNGraph(g, edata=rand(Float32, ein_channel, g.num_edges))
```
this should solve one of the test failures

melioristic commented:
I understand the error 1, this is that if my input is Float32 still my output is Float64, will see why this is happening. Could not understand the second test.

CarloLucibello commented on src/layers/conv.jl:
```diff
@@ -1065,4 +1065,119 @@ function (l::MEGNetConv)(g::GNNGraph, x::AbstractMatrix, e::AbstractMatrix)
+    (nin, ein), out = ch
+    mu = init(ein, K)
+    sigma_inv = init(ein, K)
+    b = bias ? Flux.create_bias(ones(out), true) : false
```
```suggestion
    b = bias ? Flux.create_bias(mu, true, out) : false
```
This should fix the Float64 issue

melioristic replied:
The problem is here. Both w and mu are Float32 before this, but it changes to Float64 at this point. How exactly `@.` functions? `w = @. -0.5 * (w - mu)^2`

melioristic replied:
I don't know if this is the most elegant solution, but I had to change the code at two different points. 
- `w = @. -0.5 * (w - mu)^2` goes to `w = @. - ((w - mu) ^ 2)/ 2`
- ` m = 1 / d .* m ` goes to `m = m./reshape(d, (1, g.num_nodes))`
The problem is if in degrees say d=[1,2,1] is operated to 1/d then it converts to Float 64, depending on the system. Same with 2 and 0.5 I guess.

CarloLucibello commented:
> Could not understand the second test.

Me neither. Seems to be a Zygote error due to a not supported try/catch. But we already have `@assert` and `@warn` in other layers and they don't seem to cause problems. Maybe try to simplify the forward a bit and test locally with something like
```julia
l = GMMConv((2,2)=>2)
g = rand_graph(5, 10, ndata=rand(Float32,2,5), edata=rand(Float32,2,10))
gradient(() -> sum(l(g, g.ndata.x, g.edata.e)), Flux.params(l))  
```

melioristic commented:
> > Could not understand the second test.
> 
> Me neither. Seems to be a Zygote error due to a not supported try/catch. But we already have `@assert` and `@warn` in other layers and they don't seem to cause problems. Maybe try to simplify the forward a bit and test locally with something like
> 
> ```julia
> l = GMMConv((2,2)=>2)
> g = rand_graph(5, 10, ndata=rand(Float32,2,5), edata=rand(Float32,2,10))
> gradient(() -> sum(l(g, g.ndata.x, g.edata.e)), Flux.params(l))  
> ```

It is probably related to the string interpolation error in gradient. But changing this
`"Pseudo-cordinate dim $(size(e)) does not match (ein=$(ein),num_edge=$(g.num_edges))"`
to 
`"Pseudo-cordinate dim does not match (ein, num_edge))"`
runs everything smoothly

CarloLucibello commented on src/layers/conv.jl:
```diff
@@ -1065,4 +1065,121 @@ function (l::MEGNetConv)(g::GNNGraph, x::AbstractMatrix, e::AbstractMatrix)
+
+    num_edges = g.num_edges
+    d = degree(g, dir=:in)
+    print(eltype(d))
```
```suggestion
```

CarloLucibello replied:
```suggestion
```

melioristic commented:
NANs possibly because dividing by indegree, instead changed the aggregator method to mean. This should potentially get. rid of the NAN values

CarloLucibello commented on src/layers/conv.jl:
```diff
@@ -1065,4 +1065,118 @@ function (l::MEGNetConv)(g::GNNGraph, x::AbstractMatrix, e::AbstractMatrix)
+    mu = reshape(l.mu, (ein, l.K, 1))
+    
+    w = @. ((w - mu)^2) / 2
+    w = w .* reshape(l.sigma_inv, (ein, l.K, 1))
```
```suggestion
    w = w .* reshape(l.sigma_inv.^2, (ein, l.K, 1))
```
So `sigma_inv` models an inverse standard deviation and we guarantee that the inverse variance is non-negative as it should be

CarloLucibello commented on src/layers/conv.jl:
```diff
@@ -1065,4 +1065,118 @@ function (l::MEGNetConv)(g::GNNGraph, x::AbstractMatrix, e::AbstractMatrix)
+    print(io, ", K=", l.K)
+    print(io, ", σ=", l.σ)
+    print(io, ")")
+
```
```suggestion
```

CarloLucibello commented on src/layers/conv.jl:
```diff
@@ -1065,4 +1065,118 @@ function (l::MEGNetConv)(g::GNNGraph, x::AbstractMatrix, e::AbstractMatrix)
+    (nin, ein), out = l.ch
+    print(io, "GMMConv((", nin, ",", ein, ")=>", out)
+    print(io, ", K=", l.K)
+    print(io, ", σ=", l.σ)
```
this should be printed before K and only if different from identity

melioristic replied:
So sigma should be printed before k and only if different from identity? Changing that.

CarloLucibello commented:
Great, impressive works, thanks!

melioristic commented:
> Great, impressive works, thanks!

Thanks a lot for all the help! Will try some other conv layer this weekend. :) And then will see about temporal GNN.
